Chore: Adding Public-Private Key Pair for Aggregator and Federator

simardeep1792 opened the pull request:
Description:
- Introduced public-private key pair management for secure JWT authentication between Aggregator and Federator services.
- Integrated Kubernetes Secrets to securely store and mount public keys in Aggregator and private keys in Federator.
- Updated Deployment configurations to reference the appropriate key files from the Kubernetes-mounted volume.
- Implemented JWT verification in Aggregator using the public key, ensuring only authorized Federator requests are processed.
- Modified FluxCD configuration to handle encrypted secrets using SOPS, ensuring secure key management.
- Updated Dockerfiles to include environment variables for PUBLIC_KEY_PATH in Aggregator and PRIVATE_KEY_PATH in Federator


@Stephen-ONeil  I would need your help to add the keys in Federator. Let me know when you are avaiable.

Stephen-ONeil reviewed (changes requested):
Hm, we'll either need to add key pair generation and mounting to the docker dev environment, or add a env var check for something like `IS_LOCAL_DEV` to skip all the auth related steps in `aggregator.py`, and set that env var in the docker compose.

If you want, I could take over committing to this branch to set up the federator side of things :+1: 

Alternatively, we could use a "feature flag" env var (similar to the suggested `IS_LOCAL_DEV` flag, but to be set in prod and only necessary on the header check lines) to bypass the auth in the meantime, so we can merge this branch. We could then turn the auth checks back on/remove that flag when the federator component is done as a follow up :+1: .

simardeep1792 commented:
@Stephen-ONeil , sure you can take over committing to this branch to set up the federator :)

simardeep1792 commented:
@Stephen-ONeil 

- Replaced `requests` with `httpx` for async API calls, improving performance.  
- Added `IS_LOCAL_DEV` env var to skip JWT auth in local development.  
- Optimized `Dockerfile` with multi-stage build.  
- Updated Kubernetes deployment to support `IS_LOCAL_DEV`.

Stephen-ONeil commented on aggregator/aggregator.py:
```diff
@@ -80,77 +82,75 @@
+        return None
 
-PUBLIC_KEY = load_public_key()
+PUBLIC_KEY = asyncio.run(load_public_key())
```
I'm not particularly familiar with python async code, but after a very quick skim of some docs, assuming this python module itself is called as async code (I think it would need to be for the other `async` flags to work), calling `aynscio.run` might be unsafe? It might (try to) create a second async event loop on top of the existing event loop, which could fail or create unexpected behaviour.

Let me know if you've confirmed it works as expected, again I only skimmed some of the python async docs.

Stephen-ONeil replied:
Side note, while the work of handling individual requests may benefit from asynchronous execution (so that the server isn't blocked and can keep accepting requests) the step of loading the public key is called in the top level and needs to be complete before the server can start handling requests at all, so it should either stay synchronous or be immediately `await`ed here. I believe calling it with `asyncio.run` might effectively treat it as synchronous, but that might be muddled by the possibly-creating-a-second-async-loop thing.

simardeep1792 replied:
Hey @Stephen-ONeil, I looked into it, and you're right—using asyncio.run(load_public_key()) could cause issues by creating a second event loop if the module is already running in an async context.

To avoid that, I’ve updated load_public_key() to be synchronous instead. Thanks for flagging it—I appreciate the callout!

Stephen-ONeil reviewed (commented):
Hm, I'm not too familiar with async python, so this is just a comment based on skimming some docs/familiar with async programming patterns from JS.

Stephen-ONeil commented:
@simardeep1792 heads up, I've started committing to this branch. If you have more changes to make your self, let me know so we can coordinate that smoothly :+1:

simardeep1792 commented:
@Stephen-ONeil  sure go ahead!

Stephen-ONeil commented:
Pausing my work here, prioritizing demo-visible tasks. Will return to this when able (:crossed_fingers: mid next week)

Stephen-ONeil commented on k8s/federal/sync.yaml:
```diff
@@ -123,7 +99,7 @@ spec:
   interval: 5m
   update:
     strategy: Setters
-    path: ./k8s/federal/rshiny-dashboard
```
@simardeep1792  Are these changes expected or a merging fluke on this branch? Please double check the k8s manifests in particular while reviewing!

Stephen-ONeil replied:
Meant to multi-line comment that, there's a bunch of changes in `k8s/federal/sync.yaml` that I'm unsure of here.
